node/network/gossip-support/src/lib.rs

```diff
@@ -447,6 +473,25 @@ async fn ensure_i_am_an_authority(
 	Err(util::Error::NotAValidator)
 }
 
+/// Filter out all controlled keys in the given set. Returns the number of keys removed.
+async fn remove_all_controlled(
+	keystore: &SyncCryptoStorePtr,
+	authorities: &mut Vec<AuthorityDiscoveryId>,
+) -> usize {
+	let mut to_remove = Vec::new();
+	for (i, v) in authorities.iter().enumerate() {
+		if CryptoStore::has_keys(&**keystore, &[(v.to_raw_vec(), AuthorityDiscoveryId::ID)]).await {
+			to_remove.push(i);
+		}
+	}
+
+	for i in to_remove.iter().rev().copied() {
+		authorities.remove(i);
```

Comment:
Note: this function is O(authority x keys), it could be made O(authority) in theory. In practice, number of keys should be very small (usually 1).

Reply:
Isn't this O(n_authorities)? We do a single loop over the authorities, tracking those whose keys we control in `to_remove`. Then we do another loop over `to_remove`, which has at most `n_authorities` elements. But practically, somewhere between 1 and 6 elements.

I don't think we can escape O(n_authorities) here without changing a lot about what we're doing. We'd reduce keystore calls, which probably contribute the most to the practical performance of this function, by flipping the function on its head and starting by getting all controlled `AuthorityDiscoveryId`s out of the keystore and then filtering out those from the authorities list.

Reply:
Removal from a Vec is O(n) in the worst case. But I don't think it will be a problem in practice given the amount of keys should be limited by `dispute_period`.